ipn: refactor (*LocalBackend).Start and move port polling into it.

dshynkev opened the pull request:
Besides making more sense, this also eliminates a data race wherein `controlclient.New` and `readPoller` would concurrently access `Hostinfo`, the former without any synchronization. This, together with tailscale/wireguard-go#16, should allow us to enable the `ipn` race test again (resolving tailscale/corp#255).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tailscale/tailscale/456)
<!-- Reviewable:end -->

bradfitz commented on ipn/local.go:
```diff
@@ -184,6 +179,118 @@ func (b *LocalBackend) SetDecompressor(fn func() (controlclient.Decompressor, er
+		return
+	}
+	if s == nil {
+		b.logf("weird: non-error wgengine update with status=nil: %v", s)
```
we've been using `[unexpected]` as the convention for things that shouldn't fire (instead of "weird" here)

bradfitz commented on ipn/local.go:
```diff
@@ -184,6 +179,118 @@ func (b *LocalBackend) SetDecompressor(fn func() (controlclient.Decompressor, er
 	b.newDecompressor = fn
 }
 
+func (b *LocalBackend) setClientStatus(newSt controlclient.Status) {
```
let's rename `newSt` to just `st` while you're doing this move. (I've been meaning to rename that a few times but never quite got around to it)

bradfitz replied:
add a little godoc on these methods explaining what they're called and what it does? this code's all complicated enough so let's help the next reader.

bradfitz commented on ipn/local.go:
```diff
@@ -478,8 +482,6 @@ func (b *LocalBackend) readPoller() {
 
 		b.mu.Lock()
 		if b.hiCache == nil {
-			// TODO(bradfitz): it's a little weird that this port poller
```
yay, thanks!

bradfitz commented on ipn/local.go:
```diff
@@ -478,8 +482,6 @@ func (b *LocalBackend) readPoller() {
 		if b.hiCache == nil {
-			// TODO(bradfitz): it's a little weird that this port poller
-			// is started (by NewLocalBackend) before the Start call.
 			b.hiCache = new(tailcfg.Hostinfo)
```
can you maybe document the `hiCache` field up where it's declared in the struct and say what it is?

dshynkev commented:
I changed a bit more than the feedback strictly required. For example, I have deemed `hiCache` and `netMapCache` to be confusing, especially in the absence of `hi` and `netMap`. I've renamed these to `hostinfo` and `netMap` for the names to be self-documenting. Also, I had to synchronize the logger to avoid a rare race that I saw after a few runs.

bradfitz commented on control/controlclient/direct.go:
```diff
@@ -262,6 +262,7 @@ func (c *Direct) doLogin(ctx context.Context, t *oauth2.Token, flags LoginFlags,
 	tryingNewKey := c.tryingNewKey
 	serverKey := c.serverKey
 	authKey := c.authKey
+	hostinfo := c.hostinfo
```
This is a bit misleading. This is a pointer. What's guarded by c.mu, the pointer itself, or the stuff within it?

Below we access hostinfo.BackendLogID, so hopefully that can't change concurrently.

I think we need a bit more docs on ownership rules on the hostinfo field.

dshynkev replied:
This is justified by the fact that `c.hostinfo` is owned by `c` and never modified in-place. `PollNetMap` does the same thing for the same reason. This is very much non-obvious, so I'll elaborate on that.

bradfitz commented on ipn/local.go:
```diff
@@ -63,8 +63,8 @@ type LocalBackend struct {
 	state        State
-	hiCache      *tailcfg.Hostinfo
-	netMapCache  *controlclient.NetworkMap
+	hostinfo     *tailcfg.Hostinfo
```
document whether these two pointed-to values are mutable or not once set under mu?

dshynkev commented:
I've collapsed ownership into two cases: not mutated in-place and thus locked only on pointer copy (`b.netMap`) or mutated in-place and thus locked during all operations (`b.hostinfo`, `c.hostinfo`). It seems too error-prone to have complicated semantics like "not mutated in-place except this one field, so you can avoid locking in some places". This almost didn't add any new locking except in `FakeExpireAfter`, which was broken anyway since it mutated `Expiry` with no synchronization.

bradfitz commented on ipn/local.go:
```diff
@@ -366,7 +368,7 @@ func (b *LocalBackend) Start(opts Options) error {
 		ServerURL:       b.serverURL,
 		AuthKey:         opts.AuthKey,
-		Hostinfo:        hostinfo,
+		Hostinfo:        hostinfo.Clone(),
```
Is this even safe to clone (which involves reading its fields) without the lock held?

dshynkev replied:
This is not-quite-obvious because it relies on the fact that nothing else is accessing hostinfo concurrently yet (part of why I ended up moving `readPoller` out of `NewLocalBackend`). This is also the reason for the comment below. That said, this `Clone` is actually superfluous, as I now notice that `controlclient.New` clones it again. We could just keep `b` locked, of course, to be future-proof.

dshynkev replied:
Hm, unfortunately, my `-race -count=10000` test just failed due to some very rate timeout (`-race -count=1000` succeeded several times). Let me look at that.

dshynkev replied:
Never mind, I guess the timeout was just due to the high count. Didn't expect it to actually go over 10 minutes.
